### PR TITLE
APP-4406 re-download on empty ExePath

### PR DESCRIPTION
**Pull request by abe-winter**
## What changed
- `checkNonemptyPaths` function in cloud_package_manager.go -- when a package directory is already present, this is called to decide whether certain key paths have become corrupted (missing or empty), in which case it re-downloads as if the directory were missing
- `modulesForPackage` routine in cloud package manager. This is slightly brittle: would be nice to have a canonical way to link packages and modules. Nervous also that this only works on post-replacement configs, but there's no way to check that we are getting a post-replacement config.
- take `[]Module` list in `ManagerSyncer.Sync` interface (to support package -> module lookup)
## Why
- to create a recovery path for existing machines in the wild with zero'd binaries because of power cycling
## Discussion
There is at least one edge case here: someone uploads a tarball with an empty entrypoint file. (Not missing -- empty). I don't think we check for this. RDK would keep re-downloading forever when it tries to install it.

Please comment if you can think of other cases where this check would cause bad behavior.
## Other ways to solve this problem
- hashes for tar contents. Larger project probably
- manually removing the module and re-adding it on affected machines, or releasing a new version to devices with auto-update enabled

### CODE REVIEW

File: robot/packages/cloud_package_manager.go [hunk 1]
```diff
@@ -173,8 +174,19 @@ func (m *cloudManager) Sync(ctx context.Context, packages []config.PackageConfig
...
 
+		nonEmptyPaths := make([]string, 0)
+		if p.Type == config.PackageTypeModule {
+			matchedModules := config.ModulesForPackage(p.Name, modules)
+			if len(matchedModules) == 1 {
+				nonEmptyPaths = append(nonEmptyPaths, config.RemovePlaceholderPrefix(matchedModules[0].ExePath))
+			}
+			if len(matchedModules) > 1 {
+				m.logger.Warnf("package %s matched %d > 1 modules, not doing entrypoint checking", p.Name, len(matchedModules))
+			}
```
Comment (zaporter-work) [inline robot/packages/cloud_package_manager.go:185]:
```suggestion
        if p.Type == config.PackageTypeModule {
            for _, possibleModule := range modules {
                if strings.HasPrefix(possibleModule.ExePath, p.LocalDataDirectory(m.packagesDir)){
                    nonEmptyPaths = append(nonEmptyPaths, possibleModule.ExePath)
                }
            }
        }

```

(I'd advise to just store the whole abs path, but if you want to keep your relative path logic, you could trim the prefix)

**Review by zaporter-work — CHANGES_REQUESTED** [review]
Lots of good code! A few misunderstandings about when placeholder replacement occurs but core logic is sound

File: robot/packages/cloud_package_manager.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (zaporter-work) [reply robot/packages/cloud_package_manager.go:185]: I added a suggestion below

[the remark this answers is not in the document]
Comment (zaporter-work) [reply robot/packages/cloud_package_manager.go:185]: To prove this claim, I added a 
```go
		fmt.Println("ZACK MODULES HERE FROM CLOUD_PACKAGE_MANAGER ", modules)

```

Right about here.


Output:
```
ZACK MODULES HERE FROM CLOUD_PACKAGE_MANAGER  [{zack_test_underscore /home/zack/.viam/packages/data/module/1a38b76f-cac0-451c-8f72-670994fcaf5b-test_underscore-2_5_6-linux-any/bin/module  registry zack:test_underscore map[VIAM_MODULE_ROOT:/home/zack/.viam/packages/data/module/1a38b76f-cac0-451c-8f72-670994fcaf5b-test_underscore-2_5_6-linux-any] <nil> true <nil>} {timelapse /home/zack/work/viam-timelapse/bin/module  local  map[] <nil> true <nil>}]
ZACK MODULES HERE FROM CLOUD_PACKAGE_MANAGER  [{zack_test_underscore /home/zack/.viam/packages/data/module/1a38b76f-cac0-451c-8f72-670994fcaf5b-test_underscore-2_5_6-linux-any/bin/module  registry zack:test_underscore map[VIAM_MODULE_ROOT:/home/zack/.viam/packages/data/module/1a38b76f-cac0-451c-8f72-670994fcaf5b-test_underscore-2_5_6-linux-any] <nil> true <nil>} {timelapse /home/zack/work/viam-timelapse/bin/module  local  map[] <nil> true <nil>}]
2024-03-29T20:05:16.259-0400	DEBUG	packages/cloud_package_manager.go:725	unchecked error	{"error": "remove /home/zack/.viam/packages/data/ml_model/e76d1b3b-0468-4efd-bb7f-fb1d2b352fcb-pet-feeder-2023-11-27T13-28-53: directory not empty"}
ZACK MODULES HERE FROM CLOUD_PACKAGE_MANAGER  [{zack_test_underscore /home/zack/.viam/packages/data/module/1a38b76f-cac0-451c-8f72-670994fcaf5b-test_underscore-2_5_6-linux-any/bin/module  registry zack:test_underscore map[VIAM_MODULE_ROOT:/home/zack/.viam/packages/data/module/1a38b76f-cac0-451c-8f72-670994fcaf5b-test_underscore-2_5_6-linux-any] <nil> true <nil>} {timelapse /home/zack/work/viam-timelapse/bin/module  local  map[] <nil> true <nil>}]
```

Comment (zaporter-work) [reply robot/packages/cloud_package_manager.go:185]: (you could also just keep passing the modules down to the downloadPackage -> checkNonemptyPaths fn and do the string prefix checking there)

**Comment by abe-winter** [comment]
Thanks for catch re: post-replacement config. I removed the config/replacement changes. This all happens in cloud_package_manager now, using absolute paths.

Once this seems correct to you, I'll test on a device.

This feels a little brittle -- package-to-module lookup depends on a lot of assumptions about replacement and paths. Fine for now I guess, but :grimace:.

Should I do a fast-follow PR to this one which checks for non-empty entrypoint when you upload a module? i.e. the foundEntrypoint logic in validateModuleFile [here](https://github.com/viamrobotics/rdk/blob/21dbad5f97fd0b27d88445340d70db4321a08554/cli/module_registry.go#L466).

File: robot/packages/cloud_package_manager.go [hunk 2]
```diff
@@ -202,6 +214,19 @@ func (m *cloudManager) Sync(ctx context.Context, packages []config.PackageConfig
 	return outErr
 }
 
+// modulesForPackage returns module(s) whose ExePath is in the package's directory.
+// TODO: This only works if you call it after placeholder replacement. Find a cleaner way to express this link.
```
Comment (zaporter-work) [inline robot/packages/cloud_package_manager.go:218]: Yeah -- this lack-of-link is one of the primary reasons why I've been such a big advocate of adding packages and modules the graph

File: robot/packages/cloud_package_manager.go [hunk 3]
```diff
@@ -353,11 +378,34 @@ func sanitizeURLForLogs(u string) string {
...
-		return nil
+// checkNonemptyPaths returns true if all required paths are present and non-empty.
+// This exists because we have no way to check the integrity of modules *after* they've been unpacked.
+// (We do look at checksums of downloaded tarballs, though). Once we have a better integrity check
+// system for unpacked modules, this should be removed.
+func checkNonemptyPaths(dataDir, packageName string, logger logging.Logger, paths []string) bool {
+	missingOrEmpty := 0
+	for _, nePath := range paths {
+		// note: os.Stat treats symlinks as their destination. os.Lstat would stat the link itself.
+		info, err := os.Stat(path.Join(dataDir, nePath))
```
Comment (zaporter-work) [inline robot/packages/cloud_package_manager.go:389]:
```suggestion
		info, err := os.Stat(path.Join(nePath))
```

nePath's are the module exe path which is an absolute path

File: robot/packages/cloud_package_manager.go [hunk 4]
```diff
@@ -353,11 +378,34 @@ func sanitizeURLForLogs(u string) string {
...
+		if err != nil {
+			missingOrEmpty++
+			if !os.IsNotExist(err) {
+				logger.Warnw("ignoring non-NotExist error for required path", "path", nePath, "package", packageName, "error", err.Error())
+			}
```
Comment (zaporter-work) [inline robot/packages/cloud_package_manager.go:394]: [opt/doesn't matter at all] You say 'ignoring' but we still inc missingOrEmpty so we didn't really ignore it.

Maybe `file is missing at an expected path` ??

**Review by zaporter-work — APPROVED** [review]
LGTM % that small bug with abs path and pending a manual test

File: robot/packages/cloud_package_manager.go [hunk 5]
```diff
@@ -353,11 +378,34 @@ func sanitizeURLForLogs(u string) string {
...
+		} else if info.Size() == 0 {
```
Comment (zaporter-work) [inline robot/packages/cloud_package_manager.go:395]: [opt/flyby] Are we worried about people setting their exe path to a symlink to another file in their binary? (probably not)

If we are, we might want to add a check to the filetype.

If not, feel free to resolve without comment.

File: robot/packages/cloud_package_manager.go [hunk 6]
```diff
@@ -173,8 +174,19 @@ func (m *cloudManager) Sync(ctx context.Context, packages []config.PackageConfig
...
 		m.logger.Debugf("Downloading from %s", sanitizeURLForLogs(resp.Package.Url))
 
+		nonEmptyPaths := make([]string, 0)
+		if p.Type == config.PackageTypeModule {
+			matchedModules := m.modulesForPackage(p, modules)
+			if len(matchedModules) == 1 {
+				nonEmptyPaths = append(nonEmptyPaths, matchedModules[0].ExePath)
+			}
+			if len(matchedModules) > 1 {
+				m.logger.Warnf("package %s matched %d > 1 modules, not doing entrypoint checking", p.Name, len(matchedModules))
```
Comment (benjirewis) [inline robot/packages/cloud_package_manager.go:184]:
```suggestion
				m.logger.CWarnf(ctx, "package %s matched %d > 1 modules, not doing entrypoint checking", p.Name, len(matchedModules))
```
[nit]

File: robot/impl/local_robot.go [hunk 7]
```diff
@@ -1053,7 +1053,7 @@ func (r *localRobot) Reconfigure(ctx context.Context, newConfig *config.Config)
 	// TODO(RSDK-1849): Make this non-blocking so other resources that do not require packages can run before package sync finishes.
 	// TODO(RSDK-2710) this should really use Reconfigure for the package and should allow itself to check
 	// if anything has changed.
-	err := r.packageManager.Sync(ctx, newConfig.Packages)
+	err := r.packageManager.Sync(ctx, newConfig.Packages, newConfig.Modules)
```
Comment (benjirewis) [inline robot/impl/local_robot.go:1056]: Will this cause downloads on every call to `Reconfigure` even if the user's config hasn't changed at all?

Comment (zaporter-work) [reply robot/impl/local_robot.go:1056]: Nope -- that is the point of m.managedPackages and the check if the dir exists.

Package manager is a bit special in that it stores a fair bit of state inside itself.

File: robot/packages/cloud_package_manager.go (hunk 5 above)
Comment (abe-winter) [reply robot/packages/cloud_package_manager.go:395]: that case is covered in tests -- os.Stat will follow symlinks

(`os.Lstat` will examine links directly)

File: robot/packages/cloud_package_manager.go (hunk 3 above)
Comment (zaporter-work) [reply robot/packages/cloud_package_manager.go:389]: Ah oops
```suggestion
		info, err := os.Stat(nePath)
```

That

**Review by benjirewis — APPROVED** [review]
Generally LGTM!

File: robot/packages/cloud_package_manager.go (hunk 4 above)
Comment (abe-winter) [reply robot/packages/cloud_package_manager.go]: this was a logic bug -- fixed it, good catch

File: robot/packages/cloud_package_manager.go (hunk 3 above)
Comment (abe-winter) [reply robot/packages/cloud_package_manager.go]: my test didn't cover for this + should have

added a 'doesn't re-download when file isn't corrupt' case, made sure it failed with path.Join(dataDir, absPath)

**Comment by abe-winter** [comment]
## On-device test

Version:
```
2024-04-01T23:49:24.181Z	INFO	robot_server	server/entrypoint.go:92	Viam RDK	{"git_rev":"53acb04150d2c7659502b215a0350e88fa7ec0e7"}
```

### Run 1
- fresh pi
- it downloaded the module
- it didn't re-download the module after that (i.e. no doom loop)
### Run 2
I truncated start.sh to 0. The system detected that and re-downloaded:
```
2024-04-01T23:33:11.989Z	WARN	robot_server.package_manager	packages/cloud_package_manager.go:406	a required path is empty, re-downloading	{"path":"/root/.viam/packages/data/module/6efeb7e5-91aa-4187-8ed3-e84764726078-viamboat-0_2_5-linux-arm64/./start.sh","package":"m_viam-labs_viamboat"}
```
It didn't re-download after that. ('re-download' appears only once in my shell history)
### Run 3
- after run 2, I ctrl-C'd the server and started it again
- no doom loop

**Pull request merged**